### ISSUE TITLE
Auditd wont work on newer OSs because of a default audit rule "-a never,task"

### ISSUE THREAD

**Issue by ncvicchi** [wazuh/wazuh#19399]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.5.2 | FIM | Agent | Any | openSUSE Leap 15.5 |

This issue is derived from [this issue](https://github.com/wazuh/wazuh/issues/17676) as a result of tests performed. Its origin might be related to [issue 19305](https://github.com/wazuh/wazuh/issues/19305), although this has not been confirmed, but behaviour is the same. FIM whodata tests fails, and run a realtime instead.

UPDATE: Issue #19305 is confirmed not to be related.

**Comment by ncvicchi** [comment]
Blocked by #19585

**Comment by lchico** [comment]
I'm able to make the **whodata** work, but we still have some issues with the rules that should be configured by Wazuh on audit. In order to make this work, I have to

1. Add this block of code 

```  
    <whodata>
      <startup_healthcheck>no</startup_healthcheck>
    </whodata>
    <directories whodata="yes">/var/log/testwhodata</directories>
```
2. Remove by hand the rule:

```
-a never,task
# Doing:
auditctl -d never,task
```

So, after those changes, I restart the wazuh-agent.



<details>
<summary>Result</summary>

 <img src="https://github.com/wazuh/wazuh/assets/21695385/079a0468-8367-4ad6-8862-72162e108b29">
</details>

We need to check why the rule: **-a never,task** is not being removed. We had a similar issue before [fixed here](https://github.com/wazuh/wazuh/pull/1929#:~:text=Check%20audit%20blocking%20rules%20before%20start%20whodata%20thread.%0AThe%20audit%20rule%20%2Da%20never%2Ctask%20included%20by%20default%20in%20some%20Linux%20distributions%20(Amazon%20Linux%2C%20Fedora%2029)%20blocks%20the%20audit%20output%20and%20whodata%20cannot%20read%20events.)

**Comment by ncvicchi** [comment]
Change ETA waiting for definition on which way to solve this. Issue origin is identified and a decision is pending

**Comment by jotacarma90** [comment]
Comment with info related to this issue:
https://github.com/wazuh/wazuh/issues/19585#issuecomment-1825576626

**Comment by jotacarma90** [comment]
Extending ETA after some analysis from management.

**Comment by ncvicchi** [comment]
Changed ETA waiting for Managment

**Comment by jotacarma90** [comment]
Extending ETA, due to working on another issues with high priority.

**Comment by jotacarma90** [comment]
## Update 26/12/2023
I have been researching the code related to whodata rule removal that we need to solve this problem.
Reviewing necessary steps in the issue, preparing to-do list to keep track of everything needed.
Reviewing [old PRs](https://github.com/wazuh/wazuh/pull/1929) where this issue has already been worked on.

**Comment by jotacarma90** [comment]
Passed to on hold due to highly priority issue unblocked:
https://github.com/wazuh/external-devel-requests/issues/426

**Comment by jotacarma90** [comment]
## Update 29/01/2024

I have been working on a solution to this problem. Analysing the code to remove the rule that prevents us from properly monitoring whodata on certain OSes.
Trying different ways to remove it, taking ideas from this PR where something similar was done some time ago, although it was discarded:
https://github.com/wazuh/wazuh/pull/1929

Testing locally, pending to build the necessary functionality in the configuration to decide whether or not to remove the rule in question.

**Comment by jotacarma90** [comment]
# Available options

Overview of how auditd and the "`-a never,task`" rule work:
The rule overrides the possibility of generating audit events to any process that is started when this rule is set in auditd. Therefore, in our case, once syscheck is started, there is no point in deleting this rule and restarting auditd.

Since our intention was to delete this rule based on an established FIM configuration, giving the user the choice of how to proceed in case of having any directory monitored with whodata, it is not possible to do so once the reading of the FIM configuration is done, since the syscheck process is already started at this point.
Therefore, for the correct functioning of FIM whodata, it would be necessary to act before starting the process. The options available in this case would be:

- Try to do a "pre-read" of the whodata configuration and the user who made the decision whether to remove the "`-a never,task`" rule or not, before the syscheck module starts, and at that point modify the rule. The disadvantage of this option is that we would have to take into account the centralised configuration as well as the ossec.conf file itself.

- Act on the "`-a never,task`" rule without having read any of the configuration. The option would be to leave the rule intact, but include a filter for the syscheck binary, which would allow us to continue working correctly with the whodata mode. The disadvantage is that in this case we would not be giving the user the possibility to prevent the user from modifying the audit configuration (even without even knowing if whodata is configured for that agent, since we would not have read the configuration yet):

```
-a always,task -F exe="/var/ossec/bin/wazuh-syscheckd"
-a never,task
```

**Analysing the complexity of the mentioned situations**, I wonder if we shouldn't leave the total responsibility of removing this rule to the user, and simply build a warning message mechanism to warn that when that rule set, it won't be possible to start whodata. And add a documentation note about it.

The audit documentation itself warns about such cases:
[Auditctl man](https://man7.org/linux/man-pages/man8/auditctl.8.html#top_of_page)

```
DISABLED BY DEFAULT     
       On many systems auditd is configured to install an -a never,task
       rule by default. This rule causes every new process to skip all
       audit rule processing. This is usually done to avoid a small
       performance overhead imposed by syscall auditing. If you want to
       use auditd, you need to remove that rule by deleting 10-no-
       audit.rules and adding 10-base-config.rules to the audit rules
       directory.

       If you have defined audit rules that are not matching when they
       should, check auditctl -l to make sure there is no never,task
       rule there.
```

**Comment by jotacarma90** [comment]
Issue ETA delayed due to 4.8.0 release high priority issues

**Comment by jotacarma90** [comment]
## Update 23/02/2024
I have been doing a proof of concept using a fork to see if it reloads the filter for the never,task rule, so we can avoid using the testing config for it.

**Comment by jotacarma90** [comment]
ETA delayed due to higher prio epic:
https://github.com/wazuh/wazuh/issues/21755

**Comment by MarcelKemp** [comment]
ETA delayed due to higher prio epic:
https://github.com/wazuh/wazuh/issues/21755

**Comment by jotacarma90** [comment]
ETA delayed due to higher prio epic:
https://github.com/wazuh/wazuh/issues/21755

**Comment by jotacarma90** [comment]
## Update 15/04/2024

Branch: 19399-auditd-never-task-rule
Tested simple solution to verify if we can solve the problem with the filtering rule and a fork in the wazuh-syscheckd

**Comment by MarcelKemp** [comment]
### Update
We have created the following issue to inform about the problem for `4.9.0` via the documentation: 
- Issue
So the fix to the code will be done for another version, because the tested alternatives have given problems and it is necessary to look for a different alternative that fits the product requirements.

**Comment by jotacarma90** [comment]
# Update 28/05/2024

We have had a call analyzing and investigating possible solutions and the complete functioning of this audit rule.

The key is that the rule affects those processes that we want to monitor, for example, for the nano process to generate audit events, we must remove the rule and restart the nano process.

In order for syscheck to correctly read the expected whodata (audit) alerts, the following points are necessary:
- Allow syscheck's healthcheck to be able to reconnect to the socket once it has made an attempt, to allow that when the rule is removed, and audit is restarted, it can receive the expected events.
- Remove the rule -a never,task
- Restart audit

Trying different solutions, pending to propose a solution.
